### PR TITLE
[Merged by Bors] - refactor(Data/ZMod/IntUnitsPower): generalize ZMod 2 to work for `Nat` and `Int` too

### PR DESCRIPTION
**Pull request by eric-wieser**
Instead of providing `instance : Pow ℤˣ (ZMod 2)`, this now provides `instance [Module R (Additive ℤˣ)] : Pow ℤˣ R`.
It turns out that this instance already exists for `ℕ` and `ℤ`, so all we need to do is create `instance : Module (ZMod 2) (Additive ℤˣ)` and we obtain new definitions that generalize over all three types.

As a result of the generalization, `z₂pow` has been renamed in lemmas to `uzpow` (`u`nits of `ℤ`).

The motivation here is that I need $(-1) ^ i$ to make sense (and be lawful) for `ZMod 2` (on clifford algebras) and `ℕ` (on exterior algebras).

[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Powers.20of.20.60.E2.84.A4.CB.A3.60.20by.20.60ZMod.202.60/near/397569279)

---

- [x] depends on: #7862
- [x] depends on: #7661

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/ZMod/IntUnitsPower.lean [hunk 1]
```diff
@@ -9,56 +9,107 @@ import Mathlib.Data.Int.Order.Units
...
+
+@[simp] lemma toMul_uzpow (u : Additive ℤˣ) (r : R) :
+  Additive.toMul (r • u) = Additive.toMul u ^ r := rfl
+
+@[norm_cast] lemma uzpow_natCast (u : ℤˣ) (n : ℕ) : u ^ (n : R) = u ^ n := by
+  change Additive.toMul ((n : R) • Additive.ofMul u) = _
+  rw [←nsmul_eq_smul_cast, toMul_nsmul, toMul_ofMul]
+
+lemma uzpow_ofNat (s : ℤˣ) (n : ℕ) [n.AtLeastTwo] :
+    s ^ (no_index (OfNat.ofNat n : R)) = s ^ (no_index (OfNat.ofNat n : ℕ)) :=
```
Comment (j-loreaux) [inline Mathlib/Data/ZMod/IntUnitsPower.lean:77]: Can you add a note explaining the reason for the `noindex`? Or, if it makes sense, create a library note in an appropriate location and add a link to it here.

**Review by j-loreaux — COMMENTED** [review]
This diff is very hard to follow. Would you mind trying to minimize it by ordering things the same way if possible? I don't think there is anything to complain about here, but it's hard to make sure I've checked everything.

**Comment by eric-wieser** [comment]
> Would you mind trying to minimize it by ordering things the same way if possible?

This isn't really sensible because of the new typeclass assumptions. How about I split out a PR that does everything except the titular refactor?

**Comment by eric-wieser** [comment]
Nevermind, the lemmas actually _are_ in the same order, but the git diff isn't matching them up.

**Comment by eric-wieser** [comment]
If you want a clear diff, I could rename the lemmas first?

File: Mathlib/Data/ZMod/IntUnitsPower.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Data/ZMod/IntUnitsPower.lean:78]: I added a reference to the existing library note

**Comment by eric-wieser** [comment]
> If you want a clear diff, I could rename the lemmas first? (edit: done in #8087)

Well, that was a waste of time; the diff is still hard to follow, just because the statements and proofs are changing too much!

**Comment by j-loreaux** [comment]
Thanks for doing #8087. The diff was still a bit messy, but it was just readable enough to make sense of.

bors merge
